File: lua/pay_kit/kms.lua
--[[
KMS-backed signer namespace (reserved, not implemented in v1).

The design locks in the `pay_kit.kms.*` module path now so the
post-v1 follow-up can land remote-enclave signer backends (GCP KMS,
AWS KMS, HashiCorp Vault, Turnkey, Privy, Fireblocks, HSMs) without
forcing every caller to rename their require lines.

Each factory raises `pay_kit: not implemented` so a caller who reaches
for the namespace early gets a deterministic error instead of a silent
nil. Return shape (`signer, err`) when implemented will match the
local signer family in `pay_kit.signer` (`:pubkey()`, `:sign()`,
`:fee_payer()`, `:demo()`).
]]

local M = {}

local function not_implemented(name)
  return nil, 'pay_kit.kms.' .. name .. ' is reserved for a post-v1 release; not implemented yet'
end

function M.gcp(_opts)
  return not_implemented('gcp')
end

function M.aws(_opts)
  return not_implemented('aws')
end

function M.vault(_opts)
  return not_implemented('vault')
end

return M

File: ruby/lib/pay_kit/kms.rb
# frozen_string_literal: true

require_relative "errors"

module PayKit
  # Namespace reservation for remote enclave signers (GCP KMS, AWS KMS,
  # HashiCorp Vault, etc.). The shape is locked so consumers can build
  # against `PayKit::Kms.gcp(...)` today without having to rename when
  # the actual implementations ship in a follow-up release.
  #
  # Every factory currently raises `PayKit::NotImplementedError`. Loud
  # failure is on purpose: silent fallback would mask production
  # misconfiguration (a merchant intending to sign through a managed
  # KMS service should not get a local in-process signer instead).
  #
  # When implemented, KMS signers will satisfy the same duck-type
  # contract as `PayKit::Signer::Local` (`#pubkey`, `#sign(message)`,
  # `#fee_payer?`) and add async-on-network semantics with explicit
  # `pubkey:` configuration so boot does not probe the enclave.
  module Kms
    module_function

    def gcp(key_name:, pubkey:)
      raise ::PayKit::NotImplementedError,
        "PayKit::Kms.gcp(key_name: #{key_name.inspect}, pubkey: #{pubkey.inspect}) " \
        "is reserved for a follow-up release; use PayKit::Signer.file or PayKit::Signer.env in the meantime"
    end

    def aws(key_id:, region:, pubkey:)
      raise ::PayKit::NotImplementedError,
        "PayKit::Kms.aws(key_id: #{key_id.inspect}, region: #{region.inspect}, pubkey: #{pubkey.inspect}) " \
        "is reserved for a follow-up release; use PayKit::Signer.file or PayKit::Signer.env in the meantime"
    end

    def vault(addr:, path:, pubkey:)
      raise ::PayKit::NotImplementedError,
        "PayKit::Kms.vault(addr: #{addr.inspect}, path: #{path.inspect}, pubkey: #{pubkey.inspect}) " \
        "is reserved for a follow-up release; use PayKit::Signer.file or PayKit::Signer.env in the meantime"
    end
  end
end
